universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -193,4 +208,46 @@ private long getNextBlockId(ProgramDefinition programDefinition) {
             .orElseGet(() -> 0L)
         + 1;
   }
+
+  /**
+   * Update all {@link QuestionDefinition}s in the ProgramDefinition with appropriate versions from
+   * the {@link QuestionService}.
+   */
+  private ProgramDefinition syncQuestions(ProgramDefinition programDefinition) {
+    return questionService
+        .getReadOnlyQuestionService()
+        .thenApply(
+            (roQuestionService) -> {
+              return syncQuestions(programDefinition, roQuestionService);
+            })
+        .toCompletableFuture()
+        .join();
+  }
+
+  private ProgramDefinition syncQuestions(
+      ProgramDefinition programDefinition, ReadOnlyQuestionService roQuestionService) {
+    ProgramDefinition.Builder programDefinitionBuilder = programDefinition.toBuilder();
+
+    ImmutableList.Builder<BlockDefinition> blockListBuilder = ImmutableList.builder();
+    for (BlockDefinition block : programDefinition.blockDefinitions()) {
+      BlockDefinition.Builder blockBuilder = block.toBuilder();
+
+      ImmutableList.Builder<ProgramQuestionDefinition> pqdListBuilder = ImmutableList.builder();
+      for (ProgramQuestionDefinition pqd : block.programQuestionDefinitions()) {
+        QuestionDefinition questionDefinition;
+        try {
+          questionDefinition = roQuestionService.getQuestionDefinition(pqd.id());
+        } catch (QuestionNotFoundException e) {
+          throw new RuntimeException(e);
+        }
+        pqdListBuilder.add(ProgramQuestionDefinition.create(questionDefinition));
+      }
+
+      blockBuilder.setProgramQuestionDefinitions(pqdListBuilder.build());
+      blockListBuilder.add(blockBuilder.build());
+    }
+
+    programDefinitionBuilder.setBlockDefinitions(blockListBuilder.build());
+    return programDefinitionBuilder.build();
+  }
```

Comment:
Suggestion (blocking): I think this could could be easier to read if it were broken up into smaller methods. It looks like the logic acting on a block (most of what's in the outer loop) could be extracted to a method called something like `private BlockDefinition syncBlock(BlockDefinition block, ReadOnlyQuestionService roQuestionService)`

Reply:
done

Reply:
Looks a lot better, thank you.